nav2_docking/opennav_docking/src/docking_server.cpp

```diff
@@ -424,6 +440,35 @@ void DockingServer::doInitialPerception(Dock * dock, geometry_msgs::msg::PoseSta
   }
 }
 
+void DockingServer::rotateToDock(const geometry_msgs::msg::PoseStamped & dock_pose)
+{
+  rclcpp::Rate loop_rate(controller_frequency_);
+  const double dt = 1.0 / controller_frequency_;
+  auto target_pose = dock_pose;
+  target_pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+    tf2::getYaw(target_pose.pose.orientation) + M_PI);
+
+  while (rclcpp::ok()) {
+    auto robot_pose = getRobotPoseInFrame(dock_pose.header.frame_id);
+    auto angular_distance_to_heading = angles::shortest_angular_distance(
+      tf2::getYaw(robot_pose.pose.orientation), tf2::getYaw(target_pose.pose.orientation));
+    if (fabs(angular_distance_to_heading) < rotation_angular_tolerance_) {
+      break;
+    }
+
+    auto current_vel = std::make_unique<geometry_msgs::msg::TwistStamped>();
```

Comment:
Design question: I think because we're getting updates on the dock pose orientation in the robot frame from TF, we can make the rotation closed loop if we do my suggestion in the controller of looking at hte acceleration / deceleration as a function of the distance to the goal orientation instead. That would make this open-loop and not rely on the odometry data (but closed loop in the sense that we know where we are from TF to accelerate / decelerate based off of). 

That would simplify this loop and remove the odom subscriber

Reply:
@ajtudela can you respond to this one? We have other rotate to heading behaviors in other places that do so in this way without the odom smoother

Reply:
Sorry! I forgot.

I have mixed feelings about this.

On the one hand, I like the simplicity and independence of the odometry for the docking/undocking process.

The other rotate to heading behaviors are in the regulated_pure_pursuit and rotation_shim_controller; both of these get the current angular velocity from the controller and the odom subscriber. If we remove the odom subscriber, how can we obtain the angular velocity?

On the other hand, without odometry, we would only be relying on the quality of the TF. However, angular errors could be corrected by the smooth control law.